isofit/utils/luts/channel_read.py

```diff
@@ -0,0 +1,272 @@
+"""
+Nimrod Carmon's research code as-is for comparison
+"""
+import re
+
+import numpy as np
+
+
+def how_many_bands(file_path):
+
+    num_bands = 0
+    previous_band = 0
+    number_pattern = re.compile(r"^[1-9]\d{2,}(\.\d+)?.*$")
+
+    with open(file_path, "r") as file:
+        for line in file:
+
+            line = line.strip()  # Remove leading/trailing whitespaces
+            if not line or not number_pattern.match(line):
+                continue  # Skip empty lines or lines not matching the pattern
+
+            current_band = float(
+                line.split()[0]
+            )  # Assumes numbers are in the first column
+
+            if current_band < previous_band:
+                break
+            if current_band > 300 and current_band != previous_band:
+                num_bands += 1
+                previous_band = current_band
+
+    return num_bands
+
+
+"""
+SOLZEN is in filename, needed for calculating coszen
+"""
+
+# need to complete the dict in so it replaces the self object here
+def load_chn_single(infile, multipart):
+    """Load a '.chn' output file and parse critical coefficient vectors.
+
+    These are:
+        * wl      - wavelength vector
+        * sol_irr - solar irradiance
+        * sphalb  - spherical sky albedo at surface
+        * transm  - diffuse and direct irradiance along the
+                    sun-ground-sensor path
+        * transup - transmission along the ground-sensor path only
+
+
+    If the "multipart transmittance" option is active, we will use
+    a combination of three MODTRAN runs to estimate the following
+    additional quantities:
+        * t_down_dir - direct downwelling transmittance
+        * t_down_dif - diffuse downwelling transmittance
+        * t_up_dir   - direct upwelling transmittance
+        * t_up_dif   - diffuse upwelling transmittance
+
+    !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+    Be careful with these! They are to be used only by the
+    modtran_tir functions because MODTRAN must be run with a
+    reflectivity of 1 for them to be used in the RTM defined
+    in radiative_transfer.py.
+
+    * thermal_upwelling - atmospheric path radiance
+    * thermal_downwelling - sky-integrated thermal path radiance
+        reflected off the ground and back into the sensor.
+
+    !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+    We parse them one wavelength at a time."""
+
+    multipart_transmittance = multipart
```

Comment:
We should use re instead here to find the SZA value in the filename. It would be much safer.

We can use the following:

`match = re.search(r'SZA-([\d]+\.\d+)', infile)

    if match:
        SZA = float(match.group(1))
        coszen = np.cos(np.deg2rad(SZA))`

Reply:
This has been updated to extract the solar zenith from the tp6 file and take the cosine from there:
```python
file = os.path.join(self.sim_path, self.point_to_filename(point))

solzen = self.load_tp6(f"{file}.tp6")
coszen = np.cos(solzen * np.pi / 180.0)
```